client/console/src/main/java/org/apache/syncope/client/console/audit/AuditHistoryDirectoryPanel.java

```diff
@@ -0,0 +1,316 @@
+                viewAuditHistory(modelObject, target);
+                target.add(modal);
+            }
+        }, ActionLink.ActionType.VIEW, StandardEntitlement.AUDIT_READ);
+
+        final String auditRestoreEntitlement;
+        switch (this.anyTypeKind) {
+            case USER:
+                auditRestoreEntitlement = StandardEntitlement.USER_UPDATE;
+                break;
+            case GROUP:
+                auditRestoreEntitlement = StandardEntitlement.GROUP_UPDATE;
+                break;
+            default:
+                auditRestoreEntitlement = StandardEntitlement.ANYTYPE_UPDATE;
+                break;
+        }
+
+        panel.add(new ActionLink<AnyTOAuditEntryBean>() {
+
+            private static final long serialVersionUID = -6745431735457245600L;
+
+            @Override
+            public void onClick(final AjaxRequestTarget target, final AnyTOAuditEntryBean modelObject) {
+                try {
+                    AuditHistoryDirectoryPanel.this.getTogglePanel().close(target);
+                    restore(modelObject, anyTypeKind, anyTO);
+                    target.add(container);
+                } catch (SyncopeClientException e) {
+                    LOG.error("While restoring {}", auditEntryTO.getEntityKey(), e);
+                    SyncopeConsoleSession.get().error(StringUtils.isBlank(e.getMessage())
+                        ? e.getClass().getName() : e.getMessage());
+                }
+                ((BasePage) pageRef.getPage()).getNotificationPanel().refresh(target);
+            }
+        }, ActionLink.ActionType.RESTORE, auditRestoreEntitlement);
+
+        return panel;
+    }
+
+    @Override
+    protected Collection<ActionLink.ActionType> getBatches() {
+        return Collections.emptyList();
+    }
+
+    /**
+     * Restore an object based on the audit record.
+     * <p>
+     * Note that for user objects, the original audit record masks
+     * the password and the security answer; so we cannot use the audit
+     * record to resurrect the entry based on mask data. The method behavior
+     * below will reset the audit record such that the current security answer
+     * and the password for the object are always maintained, and such properties
+     * for the user cannot be restored using audit records.
+     *
+     * @param entryBean   the entry bean
+     * @param anyTypeKind the any type kind
+     * @param anyTO       the any to
+     * @return the response
+     */
+    private ProvisioningResult<? extends AnyTO> restore(final AnyTOAuditEntryBean entryBean,
+                                                        final AnyTypeKind anyTypeKind, final AnyTO anyTO) {
+        try {
+            if (anyTypeKind == AnyTypeKind.USER) {
+                UserTO userTO = (UserTO) MAPPER.readValue(entryBean.getBefore().toString(), anyTypeKind.getTOClass());
+                UserTO current = new UserRestClient().read(anyTO.getKey());
```

Comment:
If comparison was shown against `anyTO`, why patch is generated against a freshly read object?

Reply:
In reviewing this, the problem I ran into was the mismatched eTag value for update calls. When I used the `anyTO` object's eTag value for the update, I receive an exception for mismatched eTag values. If I use a freshly read object's eTag value, then the call passes through. In looking through the codebase, I found an [example here](https://github.com/apache/syncope/blob/2_1_X/client/console/src/main/java/org/apache/syncope/client/console/panels/LinkedAccountModalPanel.java#L177-L181) that more or less does the same I think. What might be a better alternative?

Reply:
Thanks for pointing that out the existing example, which is wrong and needs to be fixed - [noted](https://issues.apache.org/jira/browse/SYNCOPE-957?focusedCommentId=16986860&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16986860).

Running into mismatched eTag value errors is something that can simply happen - possibly because someone else has just updated the same user.
From the other side, what if one thinks to restore a certain version but the effective patch sent is calculated against the latest update that someone else has just sent?

You should simply pass the `getETagValue()` from original `anyTO` as [ordinarily performed](https://github.com/apache/syncope/blob/2_1_X/client/console/src/main/java/org/apache/syncope/client/console/wizards/any/UserWizardBuilder.java#L84) during update from Admin Console.

Reply:
One additional thing you could do is to take the updated ETag from `update()`'s response and pass it to existing `anyTO`'s `setEtagValue()`.